applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py

```diff
@@ -57,19 +77,24 @@ def AnalyzeDesignAndReportToCommunicator( self, currentDesign, optimizationItera
             optimization_model_part.ProcessInfo.SetValue(KM.TIME, time_before_analysis-1)
             optimization_model_part.ProcessInfo.SetValue(KM.DELTA_TIME, 0)
 
-            response.InitializeSolutionStep()
+            # now we scope in to the directory where response operations are done
+            with IterationScope(self.model_part_controller.GetIterationDirectory(), identifier, optimizationIteration, response.IsEvaluatedInFolder()):
+                # set the optimization model part in the response
+                response.SetEvaluationModelPart(optimization_model_part)

```

Comment:
In the case of fluid optimizations, i need the updated mesh for the fluid solve. I cannot use the mesh from the optimization since I need to have time steps as well (RANS and URANS both have time steps (and of course fluid problems have different types of `elements` and `conditions` such as `FLUID_ELEMENTS`, `RANS_ELEMENTS`, `WALL_CONDITIONS`, `PERIODIC_CONDITIONS`), RANS ones are psedo time steps). So with this method it gives the opportunity to response function (in my case) to write the mdpa so, fluid simulations can read it.

I have implemented this in the following
https://github.com/KratosMultiphysics/Kratos/blob/0766630b840fb1f989e40ffc87ccdd6e5347d942/applications/RANSApplication/python_scripts/response_functions/lift_to_drag_response_function.py#L82-L95

In the case of fluid, cost of writing and reading mdpas is negligible compared to solve time of fluid primal problem.

Another reason why I write the mdpa (and final default assigned json file and other required files) is, I made the PR #8399 in such a way that, if someone wants to do further investigations with a specific iteration then they can copy the iteration and run that same iteration as a standalone simulation. This is also really important since sometimes I have to do post-processing (based on Kratos) also.

Reply:
I may extend: Until now we have assumed, that the internal responses share the optimization model part with the optimizer. This is not the case e.g. for the RANS response, since in every iteration we start an analysis (including import from an updated mdpa) from scratch. 

By passing the optimization model part, we have the chance to get the current design (shape update) from the optimizer. 

I am not completely happy with this solution yet, but we did not have a better idea until now.

Reply:
I just had another idea: 

What if the response would share the optimization_model_part with the optimizer (as it is done in the structural responses), but instead of directly using it for evaluation, it can create independent model_parts for primal/adjoint etc. 

Like that we would not use the same logic everywhere and this function would not be needed. 

what do you think about:

```python
class RANSResponse():
    def __init__(...):
        self.design_model_part = model.GetSubModelPart(...) # this is expected to be imported by the optimizer
        self.primal_mp = ... # can be different from self.design_model_part, in that case the response needs to make sure to transfer updates of the design and gradients between the modelparts
        self.adjoint_mp = ... # if needed, also different from design_model_part
```

Reply:
Or we rename the function to `def UpdateDesign(self, updated_design_mp):`, that name is more descriptive to what is actually happening.

Reply:
I have a small question regarding the proposed `RANSResponse`. How can the response know which `ModelPart` is used in the `Model` for optimization. In this case either we have to pass the name of it via `json` or via function or we need to fix it. What do you think?

I like the idea of `UpdateDesign`, its more descriptive of what we doing, and we can pass design variables as well.

Reply:
done :)

Reply:
Small change, but much better now!